### ISSUE TITLE
[strict-boolean-expressions]: New feature, this rule should additionally check value return by callbacks inside Array.some, Array.every, etc

### ISSUE THREAD

**Issue by iliubinskii** [typescript-eslint/typescript-eslint#4766]
**Repro**

```TS
[1, 2].some(x => x); // Fail

[1, 2].some(x => Boolean(x)); // Ok
```

**Expected Result**

With new rule ("strict-array-callback-return-type") should complain about non-boolean return type

**Comment by JoshuaKGoldberg** [comment]
This seems like a very specific rule tailored to a very specific use-case. Can you give us a bit more info on why you'd like to see this?

We generally avoid adding rules that target unique situations to typescript-eslint core. You can always write your own rule using the APIs described in https://typescript-eslint.io/docs/development/custom-rules.

**Comment by ilyub** [comment]
> Can you give us a bit more info on why you'd like to see this?

Sure, this rule has the same motivation as @typescript-eslint/strict-boolean-expressions rule.

Consider the following example:
```ts
interface Obj {
  value?: string;
}

function f(x: Obj): void {
  if (x.value) console.log("True");
}

function g(x: Obj[]): void {
  if (x.some(item => item.value)) console.log("True");
}
```

In the above example function "f" gets warning from "@typescript-eslint/strict-boolean-expressions":
![image](https://user-images.githubusercontent.com/70687432/160991331-50f65100-3c06-4c4a-b957-05b482942a0d.png)

The value returned from calbacks in Array.some, Array.every, etc is also used as condition.
So, it would be good to check it the same way as @typescript-eslint/strict-boolean-expressions checks condition inside "if".
So, what I suggest is that function "g" should get the same warning as function "f" for "item.value".

**Comment by ilyub** [comment]
I have changed subject of this issue, because after writing longer description I see that this should not be a new rule, but rather an additional option to @typescript-eslint/strict-boolean-expressions rule.

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/1038
